modules/shared_vpc_access/main.tf

```diff
@@ -0,0 +1,78 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+data "google_project" "service_project" {
+  project_id = var.service_project_id
+}
+
+locals {
+  gke_shared_vpc_enabled = contains(var.active_apis, "container.googleapis.com")
+  gke_s_account = local.gke_shared_vpc_enabled ? format(
+    "service-%s@container-engine-robot.iam.gserviceaccount.com",
+    data.google_project.service_project.number,
+  ) : ""
+  dataproc_shared_vpc_enabled = contains(var.active_apis, "dataproc.googleapis.com")
+  dataproc_s_account = local.dataproc_shared_vpc_enabled ? format(
+    "service-%s@dataproc-accounts.iam.gserviceaccount.com",
+    data.google_project.service_project.number
+  ) : ""
+  active_api_s_accounts = compact([local.gke_s_account, local.dataproc_s_account])
+}
+
+/******************************************
+  compute.networkUser role granted to GKE service account for GKE on shared VPC subnets
+  See: https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc
+ *****************************************/
+resource "google_compute_subnetwork_iam_member" "gke_shared_vpc_subnets" {
+  provider = google-beta
+  count    = local.gke_shared_vpc_enabled && length(var.shared_vpc_subnets) != 0 ? length(var.shared_vpc_subnets) : 0
+  subnetwork = element(
+    split("/", var.shared_vpc_subnets[count.index]),
+    index(
+      split("/", var.shared_vpc_subnets[count.index]),
+      "subnetworks",
+    ) + 1,
+  )
+  role = "roles/compute.networkUser"
+  region = element(
+    split("/", var.shared_vpc_subnets[count.index]),
+    index(split("/", var.shared_vpc_subnets[count.index]), "regions") + 1,
+  )
+  project = var.host_project_id
+  member  = format("serviceAccount:%s", local.gke_s_account)
+}
+
+/******************************************
+  container.hostServiceAgentUser role granted to GKE service account for GKE on shared VPC
+  See:https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-shared-vpc
+ *****************************************/
+resource "google_project_iam_member" "gke_host_agent" {
+  count   = local.gke_shared_vpc_enabled ? 1 : 0
+  project = var.host_project_id
+  role    = "roles/container.hostServiceAgentUser"
+  member  = format("serviceAccount:%s", local.gke_s_account)
+}
+
+/******************************************
+  compute.networkUser role granted to dataproc service account for dataproc on shared VPC subnets
+  See: https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/network#creating_a_cluster_that_uses_a_vpc_network_in_another_project
+ *****************************************/
+resource "google_project_iam_member" "dataproc_shared_vpc_network_user" {
```

Comment:
@marko7460 Shouldn't this be `google_compute_subnetwork_iam_member` to avoid granting access to all subnets in the host?

Reply:
We'd need to support both paths, so adding another option like https://github.com/terraform-google-modules/terraform-google-project-factory/pull/434/files#diff-25530a290a058747ad4e0e2f230fc886R39

Reply:
The feature was coded based on https://cloud.google.com/dataproc/docs/concepts/configuring-clusters/network#creating_a_cluster_that_uses_a_vpc_network_in_another_project. I did try giving permissions only on the subnet level and I was able to create a cluster only by specifying permissions on the subnet level. I can only confirm that those permissions are enough to create a cluster but i can't confirm if running and operating the cluster will not be an issue.

Reply:
@morgante What do you think about a future enhancement that breaks this one module into one module per API service, for the following reasons:

Consider two teams, red and blue, each wanting GKE and Dataproc.  There are 4 subnets, {red,blue}-{dataproc,gke}.  The purpose of the subnets are to prevent someone spinning up a cluster from impacting running clusters by exhausting IP space.

To support this use case a per-service module would take as input a list of subnets.  For each subnet it would manage a `google_compute_subnetwork_iam_member` resource with the service agent as a member and network user as a role.

Thoughts?